openjdk 14

jonchang opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

igas commented on Formula/openjdk.rb:
```diff
@@ -1,10 +1,8 @@
-  sha256 "01059532335fefc5e0e7a23cc79eeb1dc6fea477606981b89f259aa0e0f9abc1"
-  revision 2
+  url "https://hg.openjdk.java.net/jdk/jdk/archive/jdk-14+36.tar.bz2"
+  version "14"
```
Should we follow previous format? `14+36`

jonchang replied:
This causes an audit failure due to the version parsing code

```console
% brew audit --strict openjdk
openjdk:
  * Formula has other versions so create a versioned alias:
      cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/openjdk.rb openjdk@14+36
  * Formula has invalid versioned aliases:
      openjdk@14
```

jonchang replied:
It's also not clear to me whether we want to package patch versions of openjdk. It seems sensible to only package the *-ga versions, but looking at Debian it seems that they _do_ package patch versions so 🤷‍♂

jonchang replied:
Looks like this will need the `14.0+36` version since I don't see a way to determine the build number without a flag. Needs to be 14.0 and not just 14 due to the version parsing issue w.r.t aliases as above.

reitermarkus replied:
I'd say let's stick to only the `*-ga` versions, like AdoptOpenJDK. Maybe we should just use `14.0.0+36` for the version, since the pattern will be the same for the for the `14.0.1+x` version.

jonchang commented on Formula/openjdk.rb:
```diff
@@ -46,7 +46,8 @@ def install
 
-    libexec.install "build/macosx-x86_64-server-release/images/jdk-bundle/jdk-#{short_version}.jdk" => "openjdk.jdk"
+    jdk = Dir["build/*/images/jdk-bundle/*"].first
+    libexec.install jdk => "openjdk.jdk"
```
this is needed because `short_version` maps to `14.0.0`but it gets built as `jdk-14.jdk`

jonchang commented:
This doesn't seem right. Does `--include-test` interact poorly with `--only-dependencies`?
```
17:19:33 ==> brew install bnfc
17:19:39 
17:19:39 ==> brew install --only-dependencies bnfc
17:19:40 
17:19:40 ==> brew linkage --test bnfc
17:19:42 
17:19:42 ==> brew install --only-dependencies --include-test bnfc
17:20:40 
17:20:40 ==> brew test --verbose bnfc
17:20:41 ==> FAILED
17:20:41 Error: bnfc is missing test dependencies: ghc
```

SMillerDev commented on Formula/openjdk.rb:
```diff
@@ -17,9 +16,10 @@ class Openjdk < Formula
 
   depends_on "autoconf" => :build
 
+  # From https://jdk.java.net/archive/
```
Could we use the existing build for this?

jonchang replied:
See https://github.com/Homebrew/homebrew-core/pull/45935#issuecomment-547801045

SMillerDev commented:
I really doubt that's the only thing that breaks on a java update. Does it not build all runtime dependencies from source?

jonchang commented:
> I really doubt that's the only thing that breaks on a java update. Does it not build all runtime dependencies from source?

I think we changed this at some point. Regardless I don't think it would need to for Java? Since the jars should be more or less compatible with a newer openjdk

jonchang commented:
Once CI finishes I'm going to pull these bottles despite the `bnfc` failure since I think that's on the test bot side. Not sure why test bot isn't bringing in the test dependency.

SMillerDev commented:
> Since the jars should be more or less compatible with a newer openjdk

In my experience almost all apps will break on all major Java bumps.

jonchang commented:
Can you point me to a few pull requests where this has been the case?

SMillerDev commented:
We've never had any major bump in homebrew because it was a cask. This is just my personal experience working with java and always wanting new and shiny things.

Bo98 commented on Formula/i2p.rb:
```diff
@@ -7,7 +7,7 @@ class I2p < Formula
   bottle :unneeded
 
-  depends_on "openjdk"
+  depends_on :java => "1.8"
```
Presumably `openjdk@11` won't work here?

Either way, the `install` script below will need updating as it still points to `openjdk`.

jonchang replied:
I looked at their website and they said only java 1.8 was supported. Not sure if the test passing previously was a false positive. I'll check with openjdk@11.

Bo98 commented:
Formulae which use Gradle may be broken until 6.3 is released with OpenJDK 14 support. It's at RC4 so it should be imminent.

bayandin commented:
> Formulae which use Gradle may be broken until 6.3 is released with OpenJDK 14 support. It's at RC4 so it should be imminent.

And gradle 6.3 is coming as well https://github.com/Homebrew/homebrew-core/pull/52094

akozich commented:
> > Formulae which use Gradle may be broken until 6.3 is released with OpenJDK 14 support. It's at RC4 so it should be imminent.
> 
> And gradle 6.3 is coming as well #52094

Gradle is merged. Can we move on with java 14 now?

jonchang commented:
@akozich Yes, I noticed, that's why this branch was pushed after the gradle merge. Can you help fix some of the build failures in the CI logs? I won't have a lot of time this week to work on this.

SMillerDev commented:
Should this use the same process as python@3.8 instead?

naiba commented:
Any update?

SMillerDev commented:
> Any update?

You can see updates in the timeline, no need to ask

bayandin commented:
Fails from the last run:

Catalina (10.15)
```
==> brew audit --skip-style pdftk-java --online
==> brew fetch --retry hbase --build-bottle --force
==> brew fetch --retry tika
==> brew install --build-from-source creduce
==> brew install --build-from-source jmxtrans
==> brew install --build-from-source lumo
==> brew install --build-from-source nailgun
==> brew install --verbose --build-bottle hbase
==> brew style homebrew/core
==> brew test --verbose couchdb-lucene
==> brew test --verbose gitbucket
==> brew test --verbose grails
==> brew test --verbose hive
==> brew test --verbose micronaut
==> brew test --verbose pax-construct
==> brew test --verbose pig
==> brew test --verbose selenium-server-standalone
```

Mojave (10.14)
```
==> brew audit --skip-style pdftk-java --online
==> brew fetch --retry hbase --build-bottle --force
==> brew fetch --retry tika
==> brew install --build-from-source creduce
==> brew install --build-from-source lumo
==> brew install --build-from-source nailgun
==> brew install --verbose --build-bottle hbase
==> brew test --verbose couchdb-lucene
==> brew test --verbose gitbucket
==> brew test --verbose grails
==> brew test --verbose hive
==> brew test --verbose micronaut
==> brew test --verbose pax-construct
==> brew test --verbose pig
==> brew test --verbose selenium-server-standalone
```

High Sierra (10.13)
```
==> brew audit --skip-style pdftk-java --online
==> brew fetch --retry hbase --build-bottle --force
==> brew fetch --retry tika
==> brew install --build-from-source creduce
==> brew install --build-from-source nailgun
==> brew install --verbose --build-bottle hbase
==> brew style homebrew/core
==> brew test --verbose couchdb-lucene
==> brew test --verbose gitbucket
==> brew test --verbose grails
==> brew test --verbose hive
==> brew test --verbose micronaut
==> brew test --verbose pax-construct
==> brew test --verbose pig
==> brew test --verbose selenium-server-standalone
```


> Should this use the same process as python@3.8 instead?

Given amount of fails (~15 out of ~150 dependent on `openjdk` formulae) I would prefer to use `depends_on "openjdk@13"` and bumping `openjdk` to 14.
Since `openjdk` is already: `keg_only`  this shouldn't create such problems like `python@3.8` did

Bo98 commented:
I'd probably go for the LTS 11 rather than the EOL 13.

I have already fixed a couple of things here, namely `creduce` (#53674) and `hive` (#53675). Meanwhile, there's been updates to some formulae, including pdftk-java. Might be worth a new run to see the state of things.

SMillerDev commented:
Needs to be rebased instead of rerun though.

Bo98 commented:
Indeed, that's what I was meaning to say.

Bo98 commented:
14.0.1 is out now too.

SMillerDev commented:
In the OpenJDK source the current revision is the latest so I assume that's actually 14.0.1?

Bo98 commented:
After a major version release, they branch off into a separate repo: https://hg.openjdk.java.net/jdk-updates/jdk14u

SMillerDev commented:
```
Error: 11 failed steps!
2428
brew audit --skip-style openjdk --online
2429
brew test --verbose gitbucket
2430
brew test --verbose grails
2431
brew install --build-from-source lumo
2432
brew test --verbose micronaut
2433
brew install --build-from-source nailgun
2434
brew test --verbose pax-construct
2435
brew fetch --retry polynote
2436
brew test --verbose selenium-server-standalone
2437
brew fetch --retry hbase --build-bottle --force
2438
brew install --verbose --build-bottle hbase
```

jonchang commented:
Moved more things to `openjdk@11` and fixed a few tests. Let's see how this run goes. Might need some help debugging the maven issues if they continue.

EricFromCanada commented:
A fix for `ffmpeg2theora` is in #54297.

SMillerDev commented on Formula/gitbucket.rb:
```diff
@@ -56,10 +56,16 @@ def plist
+        exec "'#{java}' -jar #{libexec}/gitbucket.war --port=#{free_port} > output"
+      end
+      sleep 12
+    ensure
+      Process.kill(9, pid)
+      Process.wait(pid)
+    end
```
This isn't needed anymore.

SMillerDev commented on Formula/hbase.rb:
```diff
@@ -1,9 +1,9 @@
+  sha256 "ea8fa72aa6220e038e30bd7c439d181b10bd7225383f7f2d224ebb5f5397310a"
 
   bottle do
     sha256 "a177355b8ce34287e500aeabb66b78ade4ada4b529166e686dc354a4edee6830" => :catalina
```
https://hbase.apache.org/book.html#java supports JDK 11 (just has a warning in the logs)

SMillerDev replied:
And no longer seems to have a docs directory

SMillerDev commented on Formula/hbase.rb:
```diff
@@ -13,38 +13,35 @@ class Hbase < Formula
     sha256 "aa8ddbb8b3f9e1c4b8cc3523486acdb7841cd97c002a9f2959c5b320c7bb0e6c"
+
+    patch do
+      url "https://gist.githubusercontent.com/jonchang/8fea9fca3542a89c760ca5895ea240cc/raw/82728ee23f34b46d7f14fd598d3181437ebc73e8/patch.diff"
```
Can we use formula-patches if this works?

jonchang replied:
Yea, just trying to figure out CI

SMillerDev commented:
Just 4 failures remaining

```
brew test --verbose pax-construct
brew test --verbose hbase
brew install --verbose --build-bottle lumo
brew install --verbose --build-bottle opentsdb
```

jonchang commented:
- [ ] `brew test pax-construct` - need to install locally to figure this one out
- [ ] `brew test hbase` - need to figure out how to talk to the db since netcat isn't cooperating
- [ ] `brew install lumo` - I think this will need to be removed, I've done a ton of stuff and still can't figure out how to get it to build from source
- [ ] `brew install opentsdb` - will probably need more patches

Bo98 commented:
Looks like `opentsdb` doesn't support anything newer than Java 8.

SMillerDev commented:
>  Error: 9 failed steps!
brew test --verbose bnfc
brew test --verbose bnfc
brew install --build-from-source cbmc
brew install --build-from-source joshua
brew test --verbose mesos
brew install --build-from-source mgba
brew test --verbose pax-construct
brew test --verbose hbase
brew test --verbose opentsdb

SMillerDev commented:
`brew install --build-from-source cbmc` fixed
`brew install --build-from-source joshua` Source is down and the new source doesn't work
`brew install --build-from-source mgba` doesn't build right now either

jonchang commented:
Thanks Sean. I figure this has gone on for long enough, perhaps we should just merge this and then open an issue to work on the remaining failures (like for go 1.14.4). Thoughts?

SMillerDev commented:
Seems good to me

chenrui333 commented:
> Thanks Sean. I figure this has gone on for long enough, perhaps we should just merge this and then open an issue to work on the remaining failures (like for go 1.14.4). Thoughts?

I think that would be good, that would help withe upgrading the other formulae.

chenrui333 commented:
👍

SMillerDev commented:
> let's go with the merge and create a ticket to address the leftover issues.

@chenrui333 were you planning on doing this? Because it won't automerge with build failures

jonchang commented:
I'll poke at this tonight.

chenrui333 commented:
> I'll poke at this tonight.

Thanks! Let me know if you need any help.

jonchang commented:
No problem, had to fix a bug with a patch that I spotted. Assuming no other obvious problems crop up with things that are supposedly fixed I'll be able to merge this tomorrow and open an issue for the remaining failures.

chenrui333 commented:
Sounds good. We can also do the patch PR after the openjdk merge though (assuming it is not a big bug :) )

jonchang commented:
It was a merge conflict that I accidentally added so probably not good to have that sneak into Homebrew-core 😅

chenrui333 commented:
make sense 😂

SMillerDev commented:
Merging, listed the issues in #57222

chenrui333 commented:
Thanks @SMillerDev!!
